src/EditorFeatures/Core/Implementation/IntelliSense/EditorCompletionImplementation/EditorAsyncCompletionItemManager.cs

```diff
@@ -0,0 +1,661 @@
+                    return true;
+                }
+
+                // Item did not ask to be preselected.  So definitely soft select it.
+                if (item.Rules.MatchPriority == MatchPriority.Default)
+                {
+                    return true;
+                }
+            }
+
+            return false;
+        }
+
+        private IEnumerable<EditorCompletion.CompletionItemWithHighlight> GetHighlightedList(List<FilterResult> filterResults, string filterText)
+        {
+            var highlightedList = new List<EditorCompletion.CompletionItemWithHighlight>();
+            foreach (var item in filterResults)
+            {
+                var highlightedSpans = _completionHelper.GetHighlightedSpans(item.CompletionItem.FilterText, filterText, CultureInfo.CurrentCulture);
+                highlightedList.Add(new EditorCompletion.CompletionItemWithHighlight(item.CompletionItem, highlightedSpans.Select(s => s.ToSpan()).ToImmutableArray()));
+            }
+
+            return highlightedList;
+        }
+
+        private ImmutableArray<EditorCompletion.CompletionFilterWithState> GetUpdatedFilters(
+            ImmutableArray<EditorCompletion.CompletionItem> originalList,
+            List<FilterResult> filteredList,
+            ImmutableArray<EditorCompletion.CompletionFilterWithState> filters,
+            string filterText)
+        {
+            // See which filters might be enabled based on the typed code
+            var textFilteredFilters = filteredList.SelectMany(n => n.CompletionItem.Filters).Distinct();
+
+            // When no items are available for a given filter, it becomes unavailable
+            return ImmutableArray.CreateRange(filters.Select(n => n.WithAvailability(textFilteredFilters.Contains(n.Filter))));
+        }
+
+
+        private static bool IsAllPunctuation(string filterText)
+        {
+            foreach (var ch in filterText)
+            {
+                if (!char.IsPunctuation(ch))
+                {
+                    return false;
+                }
+            }
+
+            return true;
+        }
+
+        private static bool IsAfterDot(ITextSnapshot snapshot, ITrackingSpan applicableToSpan)
+        {
+            var position = applicableToSpan.GetStartPoint(snapshot).Position;
+            return position > 0 && snapshot[position - 1] == '.';
+        }
+
+        private void MakeMostRecentItem(string item)
+        {
+            var updated = false;
+
+            while (!updated)
+            {
+                var oldItems = _recentItems;
+                var newItems = oldItems.Remove(item);
+
+                if (newItems.Length == MaxMRUSize)
+                {
+                    // Remove the least recent item.
+                    newItems = newItems.RemoveAt(0);
+                }
+
+                newItems = newItems.Add(item);
+                updated = ImmutableInterlocked.InterlockedCompareExchange(ref _recentItems, newItems, oldItems) == oldItems;
```

Comment:
:bulb: `ImmutableArray<string>` is not an efficient data type for this. `ImmutableList<string>` would be better.

:bulb: If you use `ImmutableList<string>`, you can use `ImmutableInterlocked.Update` and not need to worry about the looping logic.

Reply:
Great! Thank you!

---
In reply to: [207997066](https://github.com/dotnet/roslyn/pull/29016#discussion_r207997066) [](ancestors = 207997066)

Reply:
The MRU list has 10 items in it.  There is no cost at all to using an immutable array here :)